### PR TITLE
Remove anyhow and use pure thiserror impl

### LINKED ISSUE #7: Use a std::Error based error crate like thiserror
I would recommend to use an library-level error-crate, something like [thiserror](https://github.com/dtolnay/thiserror) that works with any of the existing "Application" error systems. The upside is that users can also use something like [stable-eyre](https://crates.io/crates/stable-eyre) (which gives you backtraces on stable, works like anyhow) and other such systems.

Currently you [can't mix](https://github.com/yaahc/eyre/issues/31#issuecomment-656729031) them without additional overhead due to
```
error[E0277]: the trait bound `anyhow::Error: std::error::Error` is not satisfied
  --> src\main.rs:94:39
   |
94 |     let status = ping_c.status().await?;
   |                                       ^ the trait `std::error::Error` is not implemented 
for `anyhow::Error`
   |
   = note: required because of the requirements on the impl of `From<anyhow::Error>` for `ErrReport`
   = note: required by `from`
```

Another upside is that you don't allocate by default. (Which a user will probably box again inside their own anyhow-ish crate.)

If you're interested I can make a PR for that.

### PR DESCRIPTION
**Pull request by 0xpr03**
Simulates the anyhow<thiserror<Error>> matryoshka for contexts by
implementing this via kind-parent error types.

Closes #7

### CODE REVIEW

**Comment by 0xpr03** [comment]
Looking at the details it may be easier to use [snafu](https://docs.rs/snafu/0.6.10/snafu/) to get free optional contexts, I'm currently tried to map your context model onto thiserror, but it's [specifically noted](https://github.com/dtolnay/thiserror/issues/42) that thiserror doesn't come with such a nice .context wrapper out of the box. Snafu does support such context cases, which makes it easier to use.

The other thing is that you kind of have to use two error types to get a kind and an additional context that can be stacked infinitely from function to function. I kind of simulated this here via FooError and FooErrorKind having a transparent wrapper and a context version.

File: src/protocol.rs [hunk 1]
```diff
@@ -5,15 +5,33 @@
...
 use thiserror::Error;
 use tokio::io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};
 
 #[derive(Error, Debug)]
 pub enum ProtocolError {
-    #[error("error reading or writing data")]
-    IoError,
+    #[error(transparent)]
+    Generic(#[from] ProtocolErrorKind),
+    #[error("{}: {}",context, source)]
```
Comment (jsvana) [inline src/protocol.rs:16]: nit: can you run `rustfmt` on this?

**Review by jsvana — COMMENTED** [review]
I like this overall, but I think it might be cleaner to get rid of `Context` entirely and instead use more specific `thiserror` variants. That way we can still get the error message ("failed to read packet ID"), but we get a specific error enum variant (`ProtocolError::MalformedPacket(String)`). What do you think?

**Comment by 0xpr03** [comment]
> but we get a specific error enum variant
> ProtocolError::MalformedPacket(String)

What exactly do you want to store in the String ?

**Comment by 0xpr03** [comment]
I'd try to preserve the original error. If we're looking at the server module we have
```
.read_packet()
            .await
            .map_err(|e|ServerError::with_context(e, "failed to read response packet"))?;
```

Which internally also has 3x a context and InvalidPacketId. Which calls also `read_varint` that also has a specific `ProtocolErrorKind::InvalidVarInt` and throws for io::Error.

I think you want to preserve all of them in case of an error (or at least the underlying IO error. So you don't end up with only the top level error.
Thus I'd do something like `MalformedPacket(#[from] ProtocolError)` which contains all the others (and never loose the original error). For the stacked ProtocolErrors its a little bit harder.

Personally I [went](https://github.com/0xpr03/ts3_query/blob/d37c7aae7d33e880c779ef274ca3669744774832/src/lib.rs#L136) with snafu adding some high-level contexts in the top function and adding a Backtrace gated behind a feature, so I can backtrace errors when the blackbox of server doesn't behave as expected.

So I think its up to your liking how much detail you want to expose and where you make a cut and don't report the full error chain (not the [crate](https://crates.io/crates/error-chain) ;) ).

**Comment by jsvana** [comment]
That's a fair point. Having that context is great. I think my biggest concern here is that it seems like we're rolling context from scratch. I'm not strictly opposed to the solution you have here, but do you figure we'd have less scaffolding code if we just switched this to `snafu`?

> What exactly do you want to store in the String ?
This was just a weak compromise to have a bit of context without the extra scaffolding :)

**Comment by 0xpr03** [comment]
Snafu would allow us to get a context for free (and enforce using a context all the time if we don't add a default conversion). It's a little bit more macro heavy (some people really dislike that, up to you) and we could add an optional backtrace. I'm not sure about a nested context, but it should work.

**Comment by 0xpr03** [comment]
I could give a snafu version a run, but that'll probably have to wait until next monday.

**Pull request closed without merging**